Feature/service http client (#2)
mia-platform/custom-plugin-java: README.md

> # custom-plugin-java

Reviewer: "custom-plugin-java" -> "Mia service Java Library"

> Library that allows you to define Mia-Platform custom plugins in java easily

Reviewer: "Library that allows you to define Mia-Platform custom plugins in java easily" 
-> 
"A library  that allows you to define Mia-Platform custom services in Java easily"

## Purpose

> This plugin aims to provide an easy way to integrate the Platform.

Reviewer: "This plugin"->"This library"

## Internal Architecture

> This plugin is framework-independent. Every functionality is built using plain Java code. 

Reviewer: "This plugin"->"This library"

+ `getDirectServiceProxy`, to directly communicate with a specific microservice
    ``` java
    Service serviceClient = ServiceClientFactory.getDirectServiceProxy("my-microservice", initOptions);
    ``` 

+ `getServiceProxy`, to use microservice gateway:
     ``` java
    Service serviceClient = ServiceClientFactory.getServiceProxy(initOptions);
    ``` 
### Decorators

#### PRE decorators
##### Request
An instance of class `PreDecoratorRequest` can be used as input to the decorator handler.

The utility functions exposed by the `PreDecoratorRequest` instance can be used to access the original request:

+ `getOriginalRequestMethod()` - returns the original request method
+ `getOriginalRequestPath()` - returns the path of the original request
+ `getOriginalRequestHeaders()` - returns the headers of the original request
+ `getOriginalRequestQuery()` - returns the querystring of the original request
+ `getOriginalRequestBody()` - returns the body of the original request

In addition to the methods described above, the `PreDecoratorRequest` instance exposes an interface to modify the original request,

>  which will come forwarded by microservice-gateway to the target service. This interface is accessible using the instance method 

Reviewer: "forwarded" ->"forward"

 `changeOriginalRequest` which returns a builder for `PreDecoratorRequestProxy` object with following methods:

+ `setMethod(String newMethod)` - modify the method of the original request
+ `setPath(String newPath)` - modify the path of the original request
+ `setHeaders(Map<String, String> newHeaders)` - modify the headers of the original request
+ `setQuery (Map<String, String> newQuery)` - modify the querystring of the original request
+ `setBody(Object newBody)` - change the body of the original request

> To leave the original request unchanged, the `leaveOriginalRequestUnmodified` function can be used instead.

Reviewer: "To leave the original request unchanged, the `leaveOriginalRequestUnmodified` function can be used instead."
->
"To leave the original request unchanged, you can instead use `leaveOriginalRequestUnmodified` function"

## Internal Architecture

This plugin is framework-independent. Every functionality is built using plain Java code. 

## Usage

> To get a service proxy you can use the following method from class `ServiceCilentFactory`:

Reviewer 2: ```suggestion
To get a service proxy you can use the following method from class `ServiceClientFactory`:
```

### getServiceProxy

> Before getting your service proxy, you need to inject the appropriate InitServiceOptions instance:

Reviewer 2: ```suggestion
Before getting your service proxy, you need to inject the appropriate `InitServiceOptions` instance:
```

- Default `InitServiceOptions` (port: 3000, protocol: HTTP)
    ``` java
    InitServiceOptions initOptions = new InitServiceOptions();
    Service serviceClient = ServiceClientFactory.getServiceProxy("my-microservice", initOptions);
    ```
- Custom `InitServiceOptions`: for example, you can set custom headers (including Mia Platform headers) through the constructor's `headers` parameter:
    ``` java
    Map<String, String> headers = ... // headers

>     InitServiceOptions initOptions = new InitServiceOptions(3000, Protocol.HTTPS, headers, "");

Reviewer 2: Why the empty string?

we should provide a constructor without the last parameter if optional.

An alternative could be creating a INitServiceOptionsFactory that returns `InitServiceOptions` and allows to `.setPort/.setProtocol/.setHeaders` etc..

Author: Done, used builder to get an instance of `InitServiceOptions`

     ``` java
    Service serviceClient = ServiceClientFactory.getServiceProxy(initOptions);
    ``` 
### Decorators

#### PRE decorators
##### Request
An instance of class `PreDecoratorRequest` can be used as input to the decorator handler.

The utility functions exposed by the `PreDecoratorRequest` instance can be used to access the original request:

+ `getOriginalRequestMethod()` - returns the original request method
+ `getOriginalRequestPath()` - returns the path of the original request
+ `getOriginalRequestHeaders()` - returns the headers of the original request
+ `getOriginalRequestQuery()` - returns the querystring of the original request
+ `getOriginalRequestBody()` - returns the body of the original request

In addition to the methods described above, the `PreDecoratorRequest` instance exposes an interface to modify the original request,

>  which will come forwarded by microservice-gateway to the target service. This interface is accessible using the instance method 

Reviewer 2: ```suggestion
 which will come forwarded to the target service. This interface is accessible using the instance method 
```

#### POST decorators
##### Request
An instance of class `PostDecoratorRequest` can be used as input to the decorator handler.

The utility functions exposed by the `PostDecoratorRequest` instance can be used to access the original request:
As with the original request, the `PostDecoratorResponse` instance has useful methods for access both the original request and the original response:

+ `getOriginalRequestMethod()` - returns the original request method
+ `getOriginalRequestPath()` - returns the path of the original request
+ `getOriginalRequestHeaders()` - returns the headers of the original request
+ `getOriginalRequestQuery()` - returns the querystring of the original request
+ `getOriginalRequestBody()` - returns the body of the original request
+ `getOriginalResponseBody()` - returns the body of the original response
+ `getOriginalResponseHeaders()` - returns the headers of the original response
+ `getOriginalResponseStatusCode()` - returns the status code of the original response

In addition to the methods described above, the `PostDecoratorResponse` instance exposes an interface to modify the original response,

>  which will come forwarded by microservice-gateway to the target service. This interface is accessible using the instance method 

Reviewer 2: ```suggestion
 which will come forwarded to the original caller. This interface is accessible using the instance method 
```
